### DISCUSSION TITLE
Project status, short-term road-map (Week 1, May 2021)

### DISCUSSION THREAD

**Discussion by brainless** [covidarmy/backend#24] [General]
Hello everyone, @rgxai, @Utkarshbhimte and myself had a follow-up call after the call yesterday and chats here. Here is the summary:

1. We prefer staying serverless and focus on features since we do have time limitations
2. Twitter consumer is the project by Viksit and we add/update it (@Apoorv should have access by now, please check GitHub emails)
3. We can add more consumers, each being serverless
4. We can create an aggregation layer when needed witch a K/V storage on serverless (CloudFlare provides this among others)
5. Focus on the frontend app making a better experience
6. Possibly help other orgs consume this APIs

@SwarnimWalavalkar please share your suggested infrastructure setup.

Let us break the work into GitHub issues and take it forward.

**Comment by SwarnimWalavalkar** [comment]
## Proposed Infrastructure Setup 
- NextJS Client deployed on [Vercel](https://vercel.com/)
- Use [Vercel Serverless Functions](https://vercel.com/docs/serverless-functions/) for the API, database operations, and other auxiliary scripts
- [Upstash](https://www.upstash.com/) managed Redis database

**Reply by rgxai** [reply]
I have a few concerns and suggestions with the proposed Infrastructure Setup above

1. Earlier we were using Vercel + Mongo Atlas and ran into a connections overload problem because NextJS does not explicitly close it's connections with the database. How does a Redis Cache overcome this problem? Upstash has a limit of 1k-10k concurrent connections (which is a generous upgrade from Atlas's 500 concurrent connections) but I am not sure if those number of connections would also be enough at scale.
2. For the purpose of aggregating data from multiple sources, parsing it into a standardised format and adding more variables to it which need to be updated from time to time (such as through telegram or whatsapp bots), would redis cache be able to serve that need? Or is a node server + database approach a better option?

Let's decide accordingly.

PS in case we decide to go ahead with Redis: [This article](https://medium.com/swlh/implementing-a-simple-redis-cache-with-next-js-5103976de188) mentions some good practices on when and how to use Redis cache with NextJS

**Reply by SwarnimWalavalkar** [reply]
## Problem
- Store tweets coming in from the Twitter API on the client-side with a Redis Database
- Have that DB accessible to a WhatsApp bot

## Constraints
- 1k concurrent database connections
- 10k database commands per day

## Addersing @rgxai's concerns
In the context of the connection overload issue,
One key way Upstash can prevent connection overload is through its GraphQL API. It by default provides a GraphQL API to interface with the database. GraphQL is a simple HTTP request, hence there's no need to manage any connections.

As an additional benefit, this eliminates the need to use a separate Redis client and lets us use any graphql client to interface with the API, simplifying the development process...

About aggregating data,
We could use a Vercel serverless function to host a service that would aggregate data from all the data sources, parses it, and writes it to our datastore.

Alternatively, we can build a node server for the aggregation, and deploy it on a completely managed non-serverless option, like Digitalocean or Google's App Platform, the only reason for doing this over the serverless option could be potentially better performance and reliability...

**Reply by arn4v** [reply]
So, if I understand it right, with the proposed changes the API calls to the **Twitter API** on the client side? I.e make calls from the users device?

**Reply by rgxai** [reply]
Valid points. @brainless @Utkarshbhimte any final thoughts or comments so we can finalise and close this?

**Reply by brainless** [reply]
I just noticed one point, @SwarnimWalavalkar you mention Redis + NextJS. But this is the backend project. This is not on NextJS and we can totally control our Redis connection pool.

**Reply by brainless** [reply]
Also, we should not have any issue just running 1 large Redis server on DO. AWS has readily available Cache servers. So connection pools, management and maintenance are not blockers. The backend code, unlike say frontend code, can control connection pool.

**Reply by SwarnimWalavalkar** [reply]
@brainless True. So should we use DO or AWS to host Redis instead of Upstash?

What're your thoughts on the [GraphQL API](https://docs.upstash.com/graphql/aboutgraphqlapi) they provide? I think it can be very useful...

**Reply by arn4v** [reply]
Since we'll be using viksit/covidconnect to query the Twitter API I think we should focus on moving the business logic of the backend to NextJS API routes.

We can explore setting up our own Redis server but I believe Upstash should be good enough considering their GraphQL API.

I have an initial setup going here that you can refer to: https://github.com/covidarmy/frontend/commit/e36ccb678a33b579f2318d78b87993c428e7600c

**Reply by Utkarshbhimte** [reply]
> The backend code, unlike say frontend code, can control connection pool

@brainless are you suggesting, that we have a FE code, this BE server, and Redis DB? So we are just discussing should we replace mongo with Redis or not?

> What're your thoughts on the GraphQL API they provide? I think it can be very useful

@SwarnimWalavalkar Graphql would be helpful if we had a lot of different collections in our database. I would bring any difference in DX in near future.

**Reply by brainless** [reply]
@Utkarshbhimte no I did not mean that sorry. When I read the comment from @SwarnimWalavalkar about NextJS issues with Redis I simply pointed out we are in backend. There is no NextJS.

**Reply by arn4v** [reply]
> @SwarnimWalavalkar Graphql would be helpful if we had a lot of different collections in our database. I would bring any difference in DX in near future.

Swarnim's referring to the GraphQL API Upstash exposes to execute Redis commands. This approach would entirely avoid the connection overloading issue we have experienced with Mongo.

**Reply by brainless** [reply]
@arn4v, @SwarnimWalavalkar  I think we are over complicating. Please correct me if I am wrong, but the Mongo connection issues were from earlier `frontend` code, am I right? This is the `backend` project. Connection pools should be available.

**Comment by SwarnimWalavalkar** [comment]
# Finalized Infrastructure setup

- Frontend NextJS app & Node server backend to be deployed on [Google Cloud Run](https://cloud.google.com/run) respectively.
-  And use [Google Memorystore](https://cloud.google.com/memorystore) with Redis as the managed database

## Decision parameters

### Must have's
- A fully managed, scalable, and highly available cloud infrastructure
- A productive and hassle-free developer experience
- Be compatible with future feature needs and use-cases

## Related Issues
[Migrate from MongoDB to Redis](https://github.com/covidarmy/backend/issues/26)
[Deploy the Node backend to Cloud Run](https://github.com/covidarmy/backend/issues/27)
[Setup Backend CI/CD with Cloud Run](https://github.com/covidarmy/backend/issues/28)
[Deploy NextJS app to Cloud Run](https://github.com/covidarmy/frontend/issues/36)
[Setup Frontend CI/CD with Cloud Run](https://github.com/covidarmy/frontend/issues/37)

**Reply by rgxai** [reply]
Adding to the Must have's list which led to this decision (just for future context)

- Low latency (that's why we discarded Vercel + Upstash option due to Upstash Redis cache being in Tokyo)
- Easy Availability of developer credits so that we can reach out as a not for profit open source initiative (this seemed most easily available on Google and not so much on Heroku, Upstash or Cloudflare)
